### ISSUE TITLE
DJL Arm Support

### ISSUE THREAD

**Issue by stu1130** [deepjavalibrary/djl#375]
### Minimal Background
There are a bunch of instruction sets in the world. Two of the widely adopted are x86(32 bits)/x86_64(64 bits) from Intel and ARM. Seldom people are still using x86(32 bit) system so DJL only supports x86_64. ARM based CPU which dominates the edge device world also have a family. We will only discuss two well-known ISA in the rest of the doc. Armv7 is *32* bit only and still exists on lots of edge devices. Armv8 on the other hand, starts to support *64* bit and remains backward compatible with *32* bit system. To distinguish between 32 bit and 64 bit, people also call armv8-64 bit *aarch64* or *arm64*.

## Introduction
ARM based CPU chip has been dominating dominating the edge devices like Android, Raspberry and Nvidia Jetson. This year Apple also launch arm-based new Macbook pro & Mac Mini. C++ unlike Java “write once, run everywhere” attribute is required to recompile on the target platform. We are going to describe what DJL is missing for three engines and proposed several approaches to fill up the gap in this doc. There are many arm-based edge device. We’ll start with two widely-used platforms: *Raspberry Pi* and *Nvidia Jetson* with three major engines MXNet, PyTorch and TensorFlow.

We can then further look into the arm architecture of two platforms we focus on. Let’s look at the Raspberry Pi first. There are Raspberry Pi 3, 4 which use armv8-64 bit and older Pi 2 which is embedded with armv7-32 bit. Both of them have corresponding official [docker image support](https://github.com/docker-library/official-images#architectures-other-than-amd64). By adding armv7 and armv8 to our native package, we should be able to cover majority of Raspberry Pi and devices without GPU. Notably, ARM based GPU MALI support is out of scope as none of three engines can run against it. The only way that is closest to work is to compile the model with neo-tvm or SageMaker Neo and run against DLR engine. Customers are responsible for building the MALI-enabled DLR before we decide to offer prebuilt DLR with MALI support.

Nvidia Jetson modules is equipped with Ubuntu Aarch64 which is part of armv8. There are Jetson Nano(Maxwell), TX1(Maxwell), TX2(Pascal), Xavier NX(Volta) and AGX Xavier(Volta). All devices in the Jetson family share the same CPU ISA. The only difference is GPU architecture. Once you install Jetpack, you can run with TensorFlow, PyTorch or TensorRT

Google Coral is another popular edge device that comes with edge TPU. Although Coral CPU is Cortex-A53 which implements armv8, TPU as of today, can only be run with [TFLite](https://coral.ai/docs/edgetpu/tflite-python/) and [PyTorch](https://medium.com/pytorch/get-started-with-pytorch-cloud-tpus-and-colab-a24757b8f7fc). Most people that uses Coral heavily relies on TPU than its CPU. It could be part of *stretch goal and work on it when there is customer request.*

Besides edge devices, AWS A1 instance is powered by Arm Neoverse and provides cost-effective performance. The chip implements armv8 which means we could build one armv8 native library that runs on both Raspberry Pi 4 and AWS A1.

In particular, we are only discussing the linux-based OS as this is the only platform customer is asking for. New ARM-based macos and window are out of scope and should be added as per customer request.

## Scope of work

The tasks are divided into three pieces
* Native Library and Java C++ bridge library(JNI, JavaCpp)
* LibUtils and new package publish
* Test & Test pipeline

### 1. Native Library/ (Shared Object)
For native libraries, I propose we add following new classifiers. It aligns with [docker image support](https://github.com/docker-library/official-images#architectures-other-than-amd64)and allows us to build the binaries on a powerful machine with docker container rather than leveraging edge device that has limited CPU power and memory.

For CPU

* arm32v7
* arm64v8

and two CUDA versions support

* arm64v8-cu102
* arm64v8-cu110

#### MXNet

MXNet officially support cross-platform compilation for [Raspberry Pi](https://mxnet.apache.org/versions/1.7.0/get_started/?platform=devices&iot=raspberry-pi&) and [Nvidia Jetson](https://mxnet.apache.org/get_started/jetson_setup). In fact, MXNet CI has already built and tested those two platforms on the Jenkins that is triggered daily. We should be able to work with MXNet team to download the pre-built MXNet arm libraries directly. In terms of bridge code JNA, I can find the arm, aarch64 jar on their [jar list](https://github.com/java-native-access/jna/tree/master/lib/native) and no extra step like rebuild is needed.

#### PyTorch

Today, if you want to use PyTorch on Raspberry Pi, you will need to download a pytorch arm pip wheel from unofficial website or build from source. To name a few, [pytorch-arm-builds](https://github.com/nmilosev/pytorch-arm-builds), [pytorch-aarch64](https://github.com/Lmy0217/PyTorch-aarch64). It is risky to rely on them because those are not well-tested and hard to find the compile flags which might have significant impact on performance. Such a build requires machine and pipeline. There are several options. It could be as small as DJL team own the pipeline or as big as AWS PyTorch team official joint effort with Facebook. Once we have native libraries ready, we can take advantage of the same pipeline to rebuild the JNI on top of new native libraries.

#### TensorFlow

TensorFlow officially supports Raspberry Pi [pip wheel](https://www.tensorflow.org/install/pip). Unfortunately, we consume TensorFlow with JavaCpp bindings which is compiled along with TensorFlow. [JavaCpp itself supports arm](https://repo1.maven.org/maven2/org/bytedeco/javacpp/1.5.4/). As a result, we will have similar stories as PyTorch. We can own the pipeline and machine privately or work with sig-jvm to publish arm-based TensorFlow Java package.

### 2. LibUtils

Now we get the easier part. This section only involves LibUtils.java and build.gradle change. We are able to detect the right classifier by removing the hardcoded x86_64 [here](https://github.com/awslabs/djl/blob/e357f600ad6c9f6fd78089320ab7993f0248a267/api/src/main/java/ai/djl/util/Platform.java#L131). Most of the code are already considered for future support of arm classifier. Additionally, we’ll update the corresponding Github action or CodeBuild for native package build automation.

### 3. Test/ Test pipeline

In terms of verification, there are two steps. The first step is sanity test. At least ResNet50 & one LSTM model should be tested with new binary on new platform. As a stretch goal, all existing models in model zoo should be tested and marked if error happen. We can leverage docker image to simulate edge device but we might consider to buy the edge device to run a real test. The second steps is to capture performance benchmark data point in comparison with Python baseline. If a significant performance degradation is observed, we should file an issue and work with other team to resolve it. Next is test automation.

We can reuse the build pipeline for testing. As long as the pipeline support docker, we can do cross-platform compilation and then do testing. There are several options. It could be Github action contributors outside of Amazon can easily access. Alternatively, we can build the pipeline with CodeBuild. It have better CloudWatch integration but less accessible for 3rd contributors.

### Q & A
1. Do we still want to support 32 bit armv7?
    Yes. Arm claimed 64-bit only devices coming in 2023. All of Nvidia Jetson family are 64 bit. Now leave us only old model of Raspberry Pi. If we only decide to only support armv8, we are dropping Pi 2(armv7,released on Feb 2015) and several Zero (PCB v1.2, PCB v1.3, W). It is safe to treat those the minority. We can support armv8 first and add armv7 later if there is customer request.
2. Where is the infrastructure to test Raspberry Pi and Nvidia Jetson?
    We can leverage QEMU to emulate arm-based cpu through command-line which is able to make the whole testing pipeline automated. The pipeline can be used as part of integration test or canary test as well while not worry about the device setup. Here is a reference of how to run raspberry pi with [QEMU])https://azeria-labs.com/emulate-raspberry-pi-with-qemu/)
3. How TFLite process input without operator support?
    TfLite creates ImageProcessor for image use case in tflite-support package. As claimed by the official TensorFlow, the ultimate goal is to have a complete operator support as tf.image package. The package now is limited to android device and is working in process. The source code can be found: https://github.com/tensorflow/tflite-support/tree/master/tensorflow_lite_support/java. We can see in the future the lib will extend to IOS platform and more CV and NLP use cases.

**Comment by stu1130** [comment]
native mxnet have arm build ready https://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fedge/detail/PR-19564/4/pipeline/48/

**Comment by breandan** [comment]
Does DJL have a timeline for supporting `osx-aarch64`? I see you are publishing binaries for [`osx-x86_64`](https://search.maven.org/artifact/ai.djl.spring/djl-spring-boot-starter-mxnet-osx-x86_64/0.11/jar) and was curious about the priority of ARM and specifically other OS X targets in the near future. Thanks!

**Comment by frankfliu** [comment]
@breandan 
We are planning to add PyTorch linux-aarch64 in Q1 2022, osx-aarch64 has not planned yet.

For MXNet, if you have osx-aarch64 binary installed in the system, you should be able to use it by setting environment variable like:
```
export MXNET_LIBRARY_PATH=YOUR_PATH
```

**Comment by ccjernigan** [comment]
@frankfliu Do you have a suggestion for obtaining the osx-aarch64 binary?

I tried compiling the MxNet 1.8 branch from source on my M1 Mac, although there appears to be a number of issues making the build broken there.  The MxNet master branch does build, but it crashes at runtime if I try to use it (which seems reasonable since DJL is using MxNet 1.8 but the master branch is for v2.0)

**Comment by frankfliu** [comment]
@ccjernigan 

For M1 Mac, I recommend you try pytorch 1.11.0, what you need to do is:
1. build `libtorch` from source, manually create `libtorch` folder in pytorch-native folder, and copy required .so file and hader files into libtorch folder
2. build pytorh jni:
```
./gradlew :engines:pytorch:pytorch-native:compileJNI
```

**Comment by ccjernigan** [comment]
Thank you—I'll try that and report back how that goes!

**Comment by ccjernigan** [comment]
@frankfliu I made some progress, but I think I'm missing a step somewhere as I get a build failure with djl.
 
Here's what I did:
1. Checked out the PyTorch repo, switched to the 1.11.0 tag, and updated the submodules
2. `conda install astunparse numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing_extensions future six requests dataclasses pkg-config libuv`
3. `conda activate`
4.`export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}`
5. `USE_DISTRIBUTED=1 CMAKE_OSX_ARCHITECTURES=arm64 MACOSX_DEPLOYMENT_TARGET=11.0 USE_MKLDNN=OFF USE_NNPACK=OFF USE_QNNPACK=OFF BUILD_TEST=OFF python setup.py bdist_wheel` copied from https://github.com/pytorch/pytorch/blob/cd6363de3687a33f25848341bb4e594f7c9f99c3/.jenkins/pytorch/macos-build.sh#L36
6. Copied the outputs of `pytorch/build` to `djl/engines/pytorch/pytorch-native/libtorch`
7. Modified the build.sh to skip download
8. Ran ./gradlew :engines:pytorch:pytorch-native:compileJNI` but I get the error at the below

Besides the error message, I'll note that my build output copied in step 6 contains a somewhat different set of files compared to what comes from a fresh download of the libtorch 1.11.0 library.  I included the lists of files below as well.

Is there a step that I'm missing or something that I'm doing differently from you?

```
> Task :engines:pytorch:pytorch-native:compileJNI FAILED
+++ dirname build.sh
++ cd .
++ pwd
+ WORK_DIR=djl/engines/pytorch/pytorch-native
+ export WORK_DIR
+ NUM_PROC=1
++ command -v nproc
+ [[ -n '' ]]
++ command -v sysctl
+ [[ -n /usr/sbin/sysctl ]]
++ sysctl -n hw.ncpu
+ NUM_PROC=20
++ uname
++ tr '[:upper:]' '[:lower:]'
+ PLATFORM=darwin
+ VERSION=1.11.0
+ FLAVOR=cpu
+ AARCH64_CXX11ABI=-cxx11
+ CXX11ABI=-cxx11-abi
+ [[ cxx11 == \p\r\e\c\x\x\1\1 ]]
+ ARCH=aarch64
+ [[ ! -d libtorch ]]
+ [[ 1.11.0 =~ ^1\.10\..*|^1\.9\..* ]]
+ pushd .
~/NoSync/Development/djl/engines/pytorch/pytorch-native ~/NoSync/Development/djl/engines/pytorch/pytorch-native
+ rm -rf build
+ mkdir build
+ cd build
+ mkdir classes
+ javac -sourcepath ../../pytorch-engine/src/main/java/ ../../pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java -h include -d classes
+ cmake -DCMAKE_PREFIX_PATH=libtorch -DPT_OLD_VERSION= ..
-- The C compiler identification is AppleClang 13.1.6.13160021
-- The CXX compiler identification is AppleClang 13.1.6.13160021
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Building djl_torch with the host...
CMake Error at libtorch/Caffe2Config.cmake:14 (include):
  include could not find requested file:

    djl/engines/pytorch/pytorch-native/libtorch/public/utils.cmake
Call Stack (most recent call first):
  libtorch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:13 (find_package)


CMake Error at libtorch/Caffe2Config.cmake:17 (include):
  include could not find requested file:

    djl/engines/pytorch/pytorch-native/libtorch/public/threads.cmake
Call Stack (most recent call first):
  libtorch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:13 (find_package)


CMake Error at libtorch/Caffe2Config.cmake:109 (include):
  include could not find requested file:

    djl/engines/pytorch/pytorch-native/libtorch/public/mkl.cmake
Call Stack (most recent call first):
  libtorch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:13 (find_package)


CMake Error at libtorch/Caffe2Config.cmake:116 (include):
  include could not find requested file:

    djl/engines/pytorch/pytorch-native/libtorch/Caffe2Targets.cmake
Call Stack (most recent call first):
  libtorch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:13 (find_package)


CMake Error at libtorch/TorchConfig.cmake:186 (set_target_properties):
  set_target_properties Can not find target to add properties to: torch
Call Stack (most recent call first):
  CMakeLists.txt:13 (find_package)


-- Found Torch: djl/engines/pytorch/pytorch-native/libtorch/lib/libtorch.dylib  
-- Found JNI: /Users/carter/Library/Java/JavaVirtualMachines/azul-17.0.3/Contents/Home/lib/libjawt.dylib  
-- Configuring incomplete, errors occurred!
See also "djl/engines/pytorch/pytorch-native/build/CMakeFiles/CMakeOutput.log".
See also "djl/engines/pytorch/pytorch-native/build/CMakeFiles/CMakeError.log".
```

List of files from building pytorch from source
```
aten
bdist.macosx-10.9-x86_64
bin
build.ninja
c10
caffe2
Caffe2Config.cmake
Caffe2ConfigVersion.cmake
cmake_install.cmake
cmake_uninstall.cmake
CMakeCache.txt
CMakeFiles
compile_commands.json
confu-deps
CTestTestfile.cmake
empty.cpp
FXdiv
include
indented.txt
install_manifest.txt
lib
lib.macosx-10.9-x86_64-3.9
sleef
temp.macosx-10.9-x86_64-3.9
third_party
TorchConfig.cmake
TorchConfigVersion.cmake
```

List of files in a fresh download of pytorch binaries
```
bin
build-hash
build-version
include
lib
share
```

**Comment by ccjernigan** [comment]
I have one small update: I had better luck downloading a precompiled version of Pytorch from here https://anaconda.org/pytorch/pytorch/files?version=1.11.0.arm64

With that, I got the Gradle scripts to run in the djl project.  I think I need to make some additional changes before this fully works, but it seems like progress.

**Comment by frankfliu** [comment]
@ccjernigan 
I made a PR for M1 for macOS: https://github.com/deepjavalibrary/djl/pull/1656

**Comment by ccjernigan** [comment]
@frankfliu I tried building from your PR commit.  When I run the task `publishToMavenLocal`, I don't see `pytorch-native` appearing in my ~/.m2 repository.  Is that expected?

I do see artifacts with version 1.11.0-0.18.0-SNAPSHOT for:
 - pytorch-model-zoo
 - pytorch-engine
 - pytorch-jni

If I unzip the pytorch-jni JAR file in m2, I do see osx-aarch64 bundled inside of it as expected.

**Comment by frankfliu** [comment]
@ccjernigan 

`pytorch-native-cpu` package is option, is used for offline distribution. Without `pytorch-native-cpu`, DJL will download osx-aarch64 native library from DJL's server. I assume macOS is just for developers and always has network access, I didn't create pytorch-native-cpu for macOS M1 yet.

**Comment by frankfliu** [comment]
Now we have aarch64 support for PyTorch and OnnxRuntime.
